cmd/qat_plugin/kerneldrv/kerneldrv.go

```diff
@@ -0,0 +1,325 @@
+						fmt.Sprintf("QAT_SECTION_NAME_%s_%d", devType, uniqID): sname,
+						// This env variable may get overriden if a container requests more than one QAT process.
+						// But we keep this code since the majority of pod workloads run only one QAT process.
+						// The rest should use QAT_SECTION_NAME_XXX variables.
+						"QAT_SECTION_NAME": sname,
+					},
+				})
+				uniqID++
+			}
+
+			if !svalue.pinned {
+				break
+			}
+		}
+	}
+
+	return devTree, nil
+}
+
+// DevicePlugin represents QAT plugin exploiting kernel driver.
+type DevicePlugin struct {
+	execer    utilsexec.Interface
+	configDir string
+}
+
+// NewDevicePlugin returns new instance of kernel based QAT plugin.
+func NewDevicePlugin() *DevicePlugin {
+	return newDevicePlugin("/etc", utilsexec.New())
+}
+
+func newDevicePlugin(configDir string, execer utilsexec.Interface) *DevicePlugin {
+	return &DevicePlugin{
+		execer:    execer,
+		configDir: configDir,
+	}
+}
+
+func (dp *DevicePlugin) parseConfigs() (map[string]section, error) {
+	outputBytes, err := dp.execer.Command("adf_ctl", "status").CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Can't get driver status")
+	}
+
+	devNum, driverConfig, err := dp.getDriverConfig(string(outputBytes[:]))
+	if err != nil {
+		return nil, err
+	}
+
+	// check if the number of sections with LimitDevAccess=1 is equal to the number of endpoints
+	for sname, svalue := range driverConfig {
+		if svalue.pinned && len(svalue.endpoints) != devNum {
+			return nil, errors.Errorf("Section [%s] must be defined for all QAT devices since it contains LimitDevAccess=1", sname)
+		}
+	}
+
+	return driverConfig, nil
+}
+
+func (dp *DevicePlugin) getDriverConfig(output string) (int, map[string]section, error) {
+	devNum := 0
+	drvConfig := make(driverConfig)
+	for ln, line := range strings.Split(output, "\n") {
+		if !strings.HasPrefix(line, " qat_") {
+			continue
+		}
+
+		devstr := strings.SplitN(line, "-", 2)
+		if len(devstr) != 2 {
+			continue
+		}
+
+		devprops := strings.Split(devstr[1], ",")
+		devType := getDevType(devprops)
+
+		if devType == "" {
+			continue
+		}
+
+		devID := strings.TrimPrefix(strings.TrimSpace(devstr[0]), "qat_")
+
+		// Parse the configuration.
+		config, err := ini.Load(path.Join(dp.configDir, fmt.Sprintf("%s_%s.conf", devType, devID)))
+		if err != nil {
+			return 0, nil, errors.Wrap(err, "failed to parse device config")
+		}
+		devNum++
+
+		debug.Print(ln, devID, line)
+
+		for _, section := range config.Sections() {
+			if section.Name() == "GENERAL" || section.Name() == "KERNEL" || section.Name() == "KERNEL_QAT" || section.Name() == ini.DEFAULT_SECTION {
+				continue
+			}
+			debug.Print(section.Name())
+			if err := drvConfig.update(devID, section); err != nil {
+				return 0, nil, err
+			}
+		}
+
+	}
+
+	return devNum, drvConfig, nil
+}
+
+func getDevType(devprops []string) string {
+	devType := ""
+	for _, propstr := range devprops {
+		switch strings.TrimSpace(propstr) {
```

Comment:
also available but missing: d15xx, c3xxx

Reply:
I've removed this enumeration. Now all the device types are taken from `adf_ctl status` output.